### PR TITLE
Token filters

### LINKED ISSUE #46: Add more filters to `tokens` query
Add some filters to get tokens containing or starting with a particular string.

### PR DESCRIPTION
**Pull request by SepehrGanji**
fixing #46

### CODE REVIEW

File: src/graphql/resolvers/token-resolver.ts [hunk 1]
```diff
@@ -19,6 +19,15 @@ class TokensQueryArgs {
...
 
+  @ValidateIf((o: TokensQueryArgs) => {
+    if (o.name) {
+      const starCount = (o.name.match(/\*/g) || []).length;
+      return starCount > 2;
+    }
+
+    return false;
+  })
+  @IsEmpty({ message: "Number of stars in wildcard can't exceed 2." })
```
Comment (capt-nemo429) [inline src/graphql/resolvers/token-resolver.ts]:
```suggestion
  @IsEmpty({ message: "Number of wildcard chars can't exceed 2." })
```

File: src/context/token-repository.ts [hunk 2]
```diff
@@ -26,6 +27,12 @@ export class TokenRepository extends BaseRepository<TokenEntity> {
           tokenIds
         });
       }
+
+      if (name) {
+        const nameLike = name.replace(/\*/g, "%");
+        filterQuery = filterQuery.andWhere(`${this.alias}.name LIKE :nameLike`, { nameLike });
```
Comment (capt-nemo429) [inline src/context/token-repository.ts]: Let's optimize it a bit by checking if wildcard characters are present, if not we use equality operator, avoiding the overhead of a unnecessary `LIKE` operator

**Review by capt-nemo429 — APPROVED** [review]

**Review by capt-nemo429 — CHANGES_REQUESTED** [review]
I ended up by approving by mistake

File: src/context/token-repository.ts (hunk 2 above)
Comment (SepehrGanji) [reply src/context/token-repository.ts]: Done.

File: src/context/token-repository.ts [hunk 3]
```diff
@@ -26,6 +27,16 @@ export class TokenRepository extends BaseRepository<TokenEntity> {
           tokenIds
         });
       }
+
+      if (name) {
+        const nameLike = name.replace(/\*/g, "%");
+        if (nameLike !== name) {
```
Comment (capt-nemo429) [inline src/context/token-repository.ts]: Regular expressions are costly in terms of performance, and we are potentially doing that twice in the same request. That's not a big deal now, but we need to be prepared to growth. These little improvements summed can do a huge difference in the future with a huge load of requests.

That being said, I would suggest you to replace this verification by something like `name.indexOf("*") !== -1` and only try to replace `*` by `%` if `true`

**Review by capt-nemo429 — CHANGES_REQUESTED** [review]

File: src/context/token-repository.ts (hunk 3 above)
Comment (SepehrGanji) [reply src/context/token-repository.ts]: Done, exiting point of view!

**Review by capt-nemo429 — APPROVED** [review]

**Pull request merged**